File: credit.c
// Import libraries
#include <cs50.h>
#include <stdio.h>


int counter(long num), sum, tnum2, slnum; // temporary number and second_to_the_last number
long ccnum, fnum, lnum, tnum, f2num; // credit card, first, last, temporary, and first2 numbers

int main(void)
{
    // Prompt for card number
    ccnum = get_long("Enter card number: ");

    // Check for card length
    int count = counter(ccnum);

    //--------------------CHECKSUM--------------------------------

    //Case-01(Multiply every other digit by 2):
    tnum = ccnum ;
    while (tnum > 0)
    {
        lnum = tnum % 10 ;
        sum = sum + lnum ;
        tnum = tnum / 100 ;
    }

    // Case-02(Multiplication every other digit by 2):
    tnum = ccnum / 10 ;
    while (tnum > 0)
    {
        slnum = (tnum % 10) ;
        tnum2 = slnum * 2 ;
        sum = sum + (tnum2 % 10) + (tnum2 / 10)  ;
        tnum = tnum / 100 ;
    }

    // Check for the first, and the first two numbers
    tnum = ccnum;
    do
    {
        f2num = tnum;
        tnum = tnum / 10;
        fnum = tnum;
    }
    while (tnum > 10);

    // Checking Last number
    lnum = ccnum % 10;

    // Print AMEX, MASTERCARD, VISA, OR INVALID
    if (sum  % 10 == 0)
    {
        if (count == 16 && (f2num >= 51 && f2num <= 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((count == 13 || count == 16) && fnum == 4)
        {
            printf("VISA\n");
        }
        else if (count == 15 && (f2num == 34 || f2num == 37))
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int counter(long num)
{
    int count = 0;
    do
    {
        num /= 10;
        count++;
    }
    while (num != 0);

    return count;
}

File: credit.py
# TODO
import sys


def main():
    # Ensure correct usage of arguments
    if len(sys.argv) != 1:
        sys.exit("Usage: python credit.py")

    while True:
        try:
            # Verify input from the user
            card_number = int(input("Number: "))
            break
        except:
            continue
    # Call the Luhn’s Algorithm
    card_len, f_one, f_two, total_sums = luhn_algo(card_number)

    if total_sums == 0:
        if (card_len == 15) and (f_two == 34 or f_two == 37):
            print("AMEX")
        elif (card_len == 16) and (f_two >= 51 and f_two <= 55):
            print("MASTERCARD")
        elif (card_len == 13 or card_len == 16) and (f_one == 4):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


def luhn_algo(card_number):
    """
    Most cards use an algorithm invented by Hans Peter Luhn of IBM. According to Luhn’s algorithm,
    the validity of a credit card number can be determined (syntactically) as follows:

    1. Multiply every other digit by 2, starting with the number’s second-to-last digit,
    and then add those products’ digits together.
    2. Add the sum to the sum of the digits that weren’t multiplied by 2.
    3. If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0),
    the number is valid!

    """
    # Get card_number's length
    card_len = len(str(card_number))

    # Case-1 (multiply the other digits by 2 and get the sum)
    prod_sums = 0
    temp = card_number // 10
    while temp > 0:
        rem = temp % 10
        prod = rem * 2
        prod_sums += ((prod % 10) + (prod // 10))
        temp = temp // 100

     # Case-2 (Sum other digits that were not multiplied by 2)
    sums = 0
    temp = card_number
    while temp > 0:
        rem = temp % 10
        sums += rem
        temp = temp // 100

    # Get the total sum
    total_sums = (sums + prod_sums) % 10

    # Get the first and the first_two digits
    temp = card_number
    while temp > 10:
        f_two = temp
        temp = temp // 10
        f_one = temp

    return card_len, f_one, f_two, total_sums


if __name__ == "__main__":
    main()